### ISSUE TITLE
Background console spams errors

### ISSUE THREAD

**Issue by tophf** [Tampermonkey/tampermonkey#365]
The Tampermonkey background console unnecessarily spams errors in case prerendering/prefetching is enabled in Chrome settings and you type something in the omnibar. The cause is that a hidden tab is created during prerendering with initial tabId `-1` that replaces the current tab in-place, so its old tabId no longer exists. This can be detected using `chrome.tabs.onReplaced` event.

~~~
background.html:1 Unchecked runtime.lastError while running browserAction.setTitle: No tab with id: 623.
    at Object.setTitle (chrome-extension://gcalenpjmijncebpfijmoaglllgpjagf/rea/background.js:8:136)
    at Object.setIcon (chrome-extension://gcalenpjmijncebpfijmoaglllgpjagf/background.js:227:57)
    at g (chrome-extension://gcalenpjmijncebpfijmoaglllgpjagf/background.js:250:488)
    at chrome-extension://gcalenpjmijncebpfijmoaglllgpjagf/background.js:24:287
    at Object.f [as each] (chrome-extension://gcalenpjmijncebpfijmoaglllgpjagf/helper.js:2:66)
    at Object.commited (chrome-extension://gcalenpjmijncebpfijmoaglllgpjagf/background.js:24:265)
    at c (chrome-extension://gcalenpjmijncebpfijmoaglllgpjagf/background.js:242:223)
2017-03-10 15:18:42.776 background.html:1 Unchecked runtime.lastError while running browserAction.setBadgeText: No tab with id: 623.
    at Object.setBadgeText (chrome-extension://gcalenpjmijncebpfijmoaglllgpjagf/rea/background.js:8:202)
    at Object.setBadge (chrome-extension://gcalenpjmijncebpfijmoaglllgpjagf/background.js:228:29)
    at g (chrome-extension://gcalenpjmijncebpfijmoaglllgpjagf/background.js:250:501)
    at chrome-extension://gcalenpjmijncebpfijmoaglllgpjagf/background.js:24:287
    at Object.f [as each] (chrome-extension://gcalenpjmijncebpfijmoaglllgpjagf/helper.js:2:66)
    at Object.commited (chrome-extension://gcalenpjmijncebpfijmoaglllgpjagf/background.js:24:265)
    at c (chrome-extension://gcalenpjmijncebpfijmoaglllgpjagf/background.js:242:223)
2017-03-10 15:18:42.779 background.html:1 Unchecked runtime.lastError while running tabs.executeScript: No tab with id: 623.
    at Object.executeScript (chrome-extension://gcalenpjmijncebpfijmoaglllgpjagf/rea/background.js:3:131)
    at Object.commited (chrome-extension://gcalenpjmijncebpfijmoaglllgpjagf/background.js:24:438)
    at c (chrome-extension://gcalenpjmijncebpfijmoaglllgpjagf/background.js:242:223)
2017-03-10 15:18:43.543 background.html:1 Unchecked runtime.lastError while running browserAction.setTitle: No tab with id: 623.
    at Object.setTitle (chrome-extension://gcalenpjmijncebpfijmoaglllgpjagf/rea/background.js:8:136)
    at Object.setIcon (chrome-extension://gcalenpjmijncebpfijmoaglllgpjagf/background.js:227:57)
    at g (chrome-extension://gcalenpjmijncebpfijmoaglllgpjagf/background.js:250:488)
    at chrome-extension://gcalenpjmijncebpfijmoaglllgpjagf/background.js:24:287
    at Object.f [as each] (chrome-extension://gcalenpjmijncebpfijmoaglllgpjagf/helper.js:2:66)
    at Object.commited (chrome-extension://gcalenpjmijncebpfijmoaglllgpjagf/background.js:24:265)
    at c (chrome-extension://gcalenpjmijncebpfijmoaglllgpjagf/background.js:242:223)
2017-03-10 15:18:43.544 background.html:1 Unchecked runtime.lastError while running browserAction.setBadgeText: No tab with id: 623.
    at Object.setBadgeText (chrome-extension://gcalenpjmijncebpfijmoaglllgpjagf/rea/background.js:8:202)
    at Object.setBadge (chrome-extension://gcalenpjmijncebpfijmoaglllgpjagf/background.js:228:29)
    at g (chrome-extension://gcalenpjmijncebpfijmoaglllgpjagf/background.js:250:501)
    at chrome-extension://gcalenpjmijncebpfijmoaglllgpjagf/background.js:24:287
    at Object.f [as each] (chrome-extension://gcalenpjmijncebpfijmoaglllgpjagf/helper.js:2:66)
    at Object.commited (chrome-extension://gcalenpjmijncebpfijmoaglllgpjagf/background.js:24:265)
    at c (chrome-extension://gcalenpjmijncebpfijmoaglllgpjagf/background.js:242:223)
2017-03-10 15:18:43.545 background.html:1 Unchecked runtime.lastError while running tabs.executeScript: No tab with id: 623.
    at Object.executeScript (chrome-extension://gcalenpjmijncebpfijmoaglllgpjagf/rea/background.js:3:131)
    at Object.commited (chrome-extension://gcalenpjmijncebpfijmoaglllgpjagf/background.js:24:438)
    at c (chrome-extension://gcalenpjmijncebpfijmoaglllgpjagf/background.js:242:223)
~~~

**Comment by derjanb** [comment]
I can reproduce this at 56.0.2924.76 but not 57.0.2987.88, but I'm not sure if prefetching is simply not working here.

Actually I don't want to track and remember all tabs and their IDs to be able to tell if the tabId is valid. If I get a webNavigation.onCommitted event with a tabId then I want be able to use it and if it fails then silently ignore the error. Unfortunately this is not possible yet, at least at version 56.

See also this comment for more information:
https://bugs.chromium.org/p/chromium/issues/detail?id=366282#c7

Long story short, I'll double-check whether I can implement some improvements.

**Comment by tophf** [comment]
In Stylish we used tabs.get before setting the badge text or title:

~~~js
chrome.tabs.get(tabId, function() {
	if (chrome.runtime.lastError)
		return;
	// do something
});
~~~

**Comment by derjanb** [comment]
Probably I'll end up with something like this, but it's more or less useless overhead, that shouldn't be necessary at all. 🙄
Thanks for reporting.

**Comment by tophf** [comment]
Just in case, also happens in Chrome 59 canary so the problem didn't magically disappear by itself.

**Comment by tophf** [comment]
>fixed at beta 

I still see the same errors in v4.3.5410 when typing in the addressbar:

![clip220-fs8](https://cloud.githubusercontent.com/assets/1310400/24351025/fc2df1ce-12ec-11e7-8cfb-e166bdbd18bd.png)

**Comment by tophf** [comment]
Had to trigger a forced update in Chrome to get the new version. Looks like more than half the errors are gone, but some remain:

![clip221-fs8](https://cloud.githubusercontent.com/assets/1310400/24351157/565762fc-12ed-11e7-9722-007655c1a2d6.png)
